text/0002-observables.md

```diff
@@ -0,0 +1,472 @@
+  { done: false, description: 'mow the lawn' },
+  { done: false, description: 'do the laundry' }
+]);
+
+todos.update(draft => {
+  draft[0].done = true;
+});
+```
+
+
+#### Shiz
+
+```js
+import { value, computed } from 'shiz';
+
+const a = value(1);
+const b = computed([a], ([a]) => a * 2);
+
+function shizObservable(shiz) {
+  return readOnlyObservable(function start(set) {
+    return shiz.on('change', () => {
+      set(shiz.get());
+    });
+  }, shiz.get());
+}
+
+const observable = shizObservable(b);
+
+const unsubscribe = observable.subscribe(value => {
+  console.log(value); // logs 2
+});
+
+a.set(2); // logs 4
+```
+
+
+### Using with Sapper
+
+At present, `Store` gets privileged treatment in [Sapper](https://sapper.svelte.technology) apps. A store instance can be created per-request, for example to contain user data. This store is attached to the component tree at render time, allowing `<span>{$user.name}</span>` to be server-rendered; its data is then passed to a client-side store.
+
+It's essential that this functionality be preserved. I'm not yet sure of the best way to achieve that.
+
+> Potential corner-cases to discuss:
+> * What happens if a subscriber causes another subscriber to be removed (e.g. it results in a component subtree being destroyed)?
+> * Is `$user.name` ambiguous (i.e. is `user` the observable, or `user.name`?) and if so how do we resolve the ambiguity
+> * What happens if `$user` is declared in a scope that has an observable `user`? Do we just not subscribe?
+
+
+
+### Where does it go?
+
+This is (🐃) up for debate, but one possibility is that we put these functions in `svelte/state`:
+
+```js
+import { observable, readOnlyObservable, derivedObservable } from 'svelte/state';
+```
+
+
+## How we teach this
+
+As with RFC 1, it's crucial that this be introduced with ample demos of how the `$` prefix works, in terms of the generated code.
+
+It's arguably simpler to teach than the existing store, since it's purely concerned with data, and avoids the 'magic' of auto-attaching. We do need to pick the right terminology though, and I'm open to alternatives to 'observable'.
+
+
+## Drawbacks
+
+Like RFC 1, this is a breaking change, though RFC 1 will break existing stores anyway. The main reason not to pursue this option would be that the `$` prefix is overly magical, though I believe the convenience outweighs the modest learning curve.
+
+Another potential drawback is that anything that uses an observable (except the markup) must *itself* become observable; they are [red functions](http://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/). But this problem is presumably fundamental, rather than an avoidable consequence of the approach we've happened to choose.
+
+
+## Alternatives
+
+* The existing store (but only at the top level; there is no opportunity to add stores at a lower level in the new design)
+* Baking in an existing state management library (and its opinions)
+* Using TC39 Observables
```

Comment:
I would star faaaar away from TC39 Observables. They haven't even landed yet, so we could forget about including 'obverservable' functions for `legacy` builds. That – or offer an alternative implementation (of our own) that offers feature parity & **actually works** in most browsers.

My vote is that "alternative implementation" that actually works

Reply:
When you say 'alternative implementation' do you mean an implementation of the [proposed API](https://github.com/tc39/proposal-observable) (e.g. [zen-observable](https://github.com/zenparsing/zen-observable)? It's the API that I have my doubts about — I don't think it makes a lot of sense for the problem at hand as it permits all sorts of wacky behaviour.

Reply:
I just mean staying away from trying to integrate / incorporate existing TC39 proposal